plugin/src/main/java/org/opensearch/ml/action/undeploy/TransportUndeployModelsAction.java

```diff
@@ -0,0 +1,170 @@
+import static org.opensearch.ml.settings.MLCommonsSettings.ML_COMMONS_VALIDATE_BACKEND_ROLES;
+
+;
+
+@Log4j2
+public class TransportUndeployModelsAction extends HandledTransportAction<ActionRequest, MLUndeployModelNodesResponse> {
+    TransportService transportService;
+    ModelHelper modelHelper;
+    MLTaskManager mlTaskManager;
+    ClusterService clusterService;
+    ThreadPool threadPool;
+    Client client;
+    NamedXContentRegistry xContentRegistry;
+    DiscoveryNodeHelper nodeFilter;
+    MLTaskDispatcher mlTaskDispatcher;
+    MLModelManager mlModelManager;
+    MLStats mlStats;
+
+    private volatile boolean allowCustomDeploymentPlan;
+    private volatile boolean filterByEnabled;
+
+    @Inject
+    public TransportUndeployModelsAction(
+            TransportService transportService,
+            ActionFilters actionFilters,
+            ModelHelper modelHelper,
+            MLTaskManager mlTaskManager,
+            ClusterService clusterService,
+            ThreadPool threadPool,
+            Client client,
+            NamedXContentRegistry xContentRegistry,
+            DiscoveryNodeHelper nodeFilter,
+            MLTaskDispatcher mlTaskDispatcher,
+            MLModelManager mlModelManager,
+            MLStats mlStats,
+            Settings settings
+    ) {
+        super(MLUndeployModelsAction.NAME, transportService, actionFilters, MLDeployModelRequest::new);
+        this.transportService = transportService;
+        this.modelHelper = modelHelper;
+        this.mlTaskManager = mlTaskManager;
+        this.clusterService = clusterService;
+        this.threadPool = threadPool;
+        this.client = client;
+        this.xContentRegistry = xContentRegistry;
+        this.nodeFilter = nodeFilter;
+        this.mlTaskDispatcher = mlTaskDispatcher;
+        this.mlModelManager = mlModelManager;
+        this.mlStats = mlStats;
+        allowCustomDeploymentPlan = ML_COMMONS_ALLOW_CUSTOM_DEPLOYMENT_PLAN.get(settings);
+        filterByEnabled = ML_COMMONS_VALIDATE_BACKEND_ROLES.get(settings);
+        clusterService
+                .getClusterSettings()
+                .addSettingsUpdateConsumer(ML_COMMONS_ALLOW_CUSTOM_DEPLOYMENT_PLAN, it -> allowCustomDeploymentPlan = it);
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(ML_COMMONS_VALIDATE_BACKEND_ROLES, it -> filterByEnabled = it);
+
+    }
+
+    @Override
+    protected void doExecute(Task task, ActionRequest request, ActionListener<MLUndeployModelNodesResponse> listener) {
+        MLUndeployModelsRequest undeployModelsRequest = MLUndeployModelsRequest.fromActionRequest(request);
+        String[] modelIds = undeployModelsRequest.getModelIds();
+        String[] targetNodeIds = undeployModelsRequest.getNodeIds();
+        boolean specifiedModelIds = modelIds != null && modelIds.length > 0;
+        modelIds = specifiedModelIds ? modelIds : mlModelManager.getAllModelIds();
+        Set<String> invalidAccessModels = new HashSet<>();
```

Comment:
HashSet is not thread safe

Reply:
Oh got it. Will look into other alternatives then. Thanks